### PR TITLE
Use sync IO when the physical path is available

### PR DESCRIPTION
**Pull request by davidfowl**
- This should remove the overhead of doing async io for the synchronous configuration system

Found this when looking at a memory profile:

image

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Configuration.FileExtensions/src/FileConfigurationProvider.cs [hunk 1]
```diff
@@ -80,16 +82,35 @@ private void Load(bool reload)
...
                 {
-                    try
+                    if (fileInfo.PhysicalPath != null)
                     {
-                        Load(stream);
-                    }
-                    catch (Exception e)
-                    {
-                        HandleException(ExceptionDispatchInfo.Capture(e));
+                        // The default physical file info assumes asynchronous IO which results in unnecessary overload
```
Comment (danmosemsft) [inline src/libraries/Microsoft.Extensions.Configuration.FileExtensions/src/FileConfigurationProvider.cs:90]: Overhead?

File: src/libraries/Microsoft.Extensions.Configuration.FileExtensions/src/FileConfigurationProvider.cs [hunk 2]
```diff
@@ -80,16 +82,35 @@ private void Load(bool reload)
...
+                            FileShare.ReadWrite,
+                            bufferSize: 1,
+                            FileOptions.SequentialScan);
                     }
+
+                    return fileInfo.CreateReadStream();
+                }
+
+                using Stream stream = OpenRead(file);
+
```
Comment (HaoK) [inline src/libraries/Microsoft.Extensions.Configuration.FileExtensions/src/FileConfigurationProvider.cs:106]: nit: extra line?

File: src/libraries/Microsoft.Extensions.Configuration.FileExtensions/src/FileConfigurationProvider.cs [hunk 3]
```diff
@@ -80,16 +82,35 @@ private void Load(bool reload)
                 {
                     Data = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
                 }
-                using (var stream = file.CreateReadStream())
+
+                static Stream OpenRead(IFileInfo fileInfo)
```
Comment (HaoK) [inline src/libraries/Microsoft.Extensions.Configuration.FileExtensions/src/FileConfigurationProvider.cs:86]: Maybe this is a stylistic thing, but is OpenRead simple enough to just inline?

```
// The default physical file info assumes asynchronous IO which results in unnecessary overload
using Stream stream = (file.PhysicalPath != null) ? new FileStream(...) : file.CreateReadStream()
```

**Review by HaoK — APPROVED** [review]

File: src/libraries/Microsoft.Extensions.Configuration.FileExtensions/src/FileConfigurationProvider.cs (hunk 1 above)
Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.Configuration.FileExtensions/src/FileConfigurationProvider.cs:90]: Oops

File: src/libraries/Microsoft.Extensions.Configuration.FileExtensions/src/FileConfigurationProvider.cs (hunk 3 above)
Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.Configuration.FileExtensions/src/FileConfigurationProvider.cs:86]: too many arguments to the FileStream ctor. I had this at first.

**Review by maryamariyan — APPROVED** [review]

**Comment by davidfowl** [comment]
Should I worry about dotnet-runtime-perf?

**Comment by maryamariyan** [comment]
> Should I worry about dotnet-runtime-perf?

@safern reruning it failed again. 

I don't seem to find extra information on this: 
https://dev.azure.com/dnceng/public/_build/results?buildId=686076&view=ms.vss-test-web.build-test-results-tab&runId=21312492&paneView=attachments&resultId=100015

**Comment by maryamariyan** [comment]
> Should I worry about dotnet-runtime-perf?

@davidfowl issue is reported here already, so should be OK to merge: https://github.com/dotnet/runtime/issues/37732.

**Comment by ooooolivia** [comment]
@davidfowl  The issue in dotnet-runtime-perf was fixed on 6/13 and your build should be able to pass if you rerun it.

**Comment by davidfowl** [comment]
Failed again?

**Comment by ooooolivia** [comment]
https://github.com/dotnet/runtime/pull/37957 This PR seems to just get merged to disable the test that just failed.

**Pull request merged**